File: src/engine/scoring.ts
/**
 * Scoring and level advancement for 升级.
 *
 * Based on defender's points at end of round:
 *   0 pts           → Declarer +3 levels
 *   5-35 pts        → Declarer +2 levels
 *   40-75 pts       → Declarer +1 level
 *   80-115 pts      → Switch: defender becomes next declarer
 *   120-155 pts     → Defender +1 level
 *   160-195 pts     → Defender +2 levels
 *   200 pts         → Defender +3 levels
 *
 * 扣底 (Ambush): If defender wins the last trick,
 *   bottom card points are doubled (×2 for singles, ×4 for pairs/tractors).
 */

import { LEVELS } from '../core/constants';
import { Rank } from '../core/card';
import type { Card } from '../core/card';

/** Result of scoring a round. */
export interface ScoreResult {
  /** How many levels the declarer team changes (>0 = up, <0 = down). */
  declarerLevelChange: number;
  /** Which team will be the next declarer (0 or 1). */
  nextDeclarerTeam: number;
  /** The next level to play. */
  nextLevel: Rank;
  /** The current team's new level. */
  team0NewLevel: Rank;
  team1NewLevel: Rank;
  /** Points breakdown. */
  defenderPoints: number;
  bottomCardBonus: number;
  totalDefenderPoints: number;
}

/**
 * Calculate the score result for a round.
 *
 * @param defenderPoints - Points collected by defender from tricks.
 * @param bottomCards - The 8 bottom cards (底牌).
 * @param lastTrickWinnerTeam - Which team won the last trick (0=declarer, 1=defender).
 * @param declarerTeamIndex - Which team was declarer this round.
 * @param currentLevel - The level being played.
 */
export function calculateScore(
  defenderPoints: number,
  bottomCards: Card[],
  lastTrickWinnerTeam: number,
  declarerTeamIndex: number,
  currentLevel: Rank,
): ScoreResult {
  const defenderTeamIndex = declarerTeamIndex === 0 ? 1 : 0;

  // Calculate bottom card bonus (扣底)
  let bottomCardPoints = 0;
  for (const card of bottomCards) {
    bottomCardPoints += card.points;
  }

  // Ambush multiplier: ×2 if defender wins last trick
  // ×4 if the winning play was a pair/tractor (simplified: ×2 for now)
  const ambushMultiplier = lastTrickWinnerTeam === defenderTeamIndex ? 2 : 0;
  const bottomCardBonus = bottomCardPoints * ambushMultiplier;

  const totalPoints = defenderPoints + bottomCardBonus;

  // Determine level change
  let declarerChange: number;
  let switchDeclarer: boolean;

  if (totalPoints === 0) {
    declarerChange = 3;
    switchDeclarer = false;
  } else if (totalPoints < 40) {
    declarerChange = 2;
    switchDeclarer = false;
  } else if (totalPoints < 80) {
    declarerChange = 1;
    switchDeclarer = false;
  } else if (totalPoints < 120) {
    declarerChange = 0;
    switchDeclarer = true;
  } else if (totalPoints < 160) {
    declarerChange = -1;
    switchDeclarer = true;
  } else if (totalPoints < 200) {
    declarerChange = -2;
    switchDeclarer = true;
  } else {
    // 200 points
    declarerChange = -3;
    switchDeclarer = true;
  }

  // Calculate new levels
  const currentLevelIndex = LEVELS.indexOf(currentLevel);
  const declarerTeamIndex_actual = declarerTeamIndex;

  let team0Change = 0;
  let team1Change = 0;

  if (declarerTeamIndex_actual === 0) {
    team0Change = declarerChange;
    team1Change = declarerChange < 0 ? -declarerChange : 0;
  } else {
    team1Change = declarerChange;
    team0Change = declarerChange < 0 ? -declarerChange : 0;
  }

  // But levels only change for the declarer side if positive,
  // and for the defender side if the declarer change is negative

  // Simplified: just track the level change
  const team0LevelIndex = Math.max(0, Math.min(LEVELS.length - 1, currentLevelIndex + team0Change));
  const team1LevelIndex = Math.max(0, Math.min(LEVELS.length - 1, currentLevelIndex + team1Change));

  const nextDeclarerTeam = switchDeclarer ? defenderTeamIndex : declarerTeamIndex;
  const nextLevel = nextDeclarerTeam === 0 ? LEVELS[team0LevelIndex] : LEVELS[team1LevelIndex];

  return {
    declarerLevelChange: declarerChange,
    nextDeclarerTeam,
    nextLevel,
    team0NewLevel: LEVELS[team0LevelIndex],
    team1NewLevel: LEVELS[team1LevelIndex],
    defenderPoints,
    bottomCardBonus,
    totalDefenderPoints: totalPoints,
  };
}

/**
 * Check if a team has reached the target level (game over).
 */
export function isGameOver(level: Rank, targetLevel: Rank): boolean {
  const levelIndex = LEVELS.indexOf(level);
  const targetIndex = LEVELS.indexOf(targetLevel);
  return levelIndex > targetIndex; // Must pass target (e.g., reach A and win one more round)
}

File: server/engine/scoring.py
"""Scoring and level advancement for 升级 (Shengji/Tractor).

Handles round scoring calculation, level advancement, and scoring/level constants.
Ported from src/engine/scoring.ts with Bug #4 and Bug #6 fixes.

Bug #4: Ambush multiplier is now SINGLE x2, PAIR x4, TRACTOR x8 (was fixed x2).
Bug #6: is_game_over uses >= (game ends when reaching A, not passing A).
"""

from dataclasses import dataclass

from pydantic import BaseModel, ConfigDict

from server.engine.card import Card, Rank
from server.engine.card_utils import POINTS_MAP
from server.engine.constants import LEVELS
from server.engine.types import PlayType


# ---- Constants ----

TARGET_LEVEL: Rank = Rank.ACE


@dataclass(frozen=True)
class ScoreThreshold:
    """One row of the scoring table."""
    max_points: int
    declarer_change: int
    switch_declarer: bool


SCORE_TABLE: list[ScoreThreshold] = [
    ScoreThreshold(max_points=0,   declarer_change=3,  switch_declarer=False),
    ScoreThreshold(max_points=40,  declarer_change=2,  switch_declarer=False),
    ScoreThreshold(max_points=80,  declarer_change=1,  switch_declarer=False),
    ScoreThreshold(max_points=120, declarer_change=0,  switch_declarer=True),
    ScoreThreshold(max_points=160, declarer_change=-1, switch_declarer=True),
    ScoreThreshold(max_points=200, declarer_change=-2, switch_declarer=True),
]

DEFAULT_SETTINGS: dict = {
    "model": "gpt-4o",
    "base_url": "https://api.openai.com/v1",
    "target_level": Rank.ACE,
    "bottom_card_count": 8,
}


# ---- Result Models ----


class ScoreResult(BaseModel):
    """Result of scoring a round."""

    model_config = ConfigDict(frozen=True)

    declarer_level_change: int
    switch_declarer: bool
    bottom_card_bonus: int
    total_defender_points: int
    team0_new_level: Rank
    team1_new_level: Rank


# ---- Scoring Logic ----


_AMBUSH_MULTIPLIER: dict[PlayType, int] = {
    PlayType.SINGLE: 2,
    PlayType.PAIR: 4,
    PlayType.TRACTOR: 8,
    PlayType.THROW: 2,
}


def _bottom_card_points(bottom_cards: list[Card]) -> int:
    """Sum point values of bottom cards using POINTS_MAP."""
    return sum(POINTS_MAP[card.rank] for card in bottom_cards)


def _determine_level_change(total_points: int) -> tuple[int, bool]:
    """Return (declarer_level_change, switch_declarer) from total defender points.

    Scoring table:
      0 pts           → declarer +3
      1-39 pts        → declarer +2
      40-79 pts       → declarer +1
      80-119 pts      → switch declarer
      120-159 pts     → defender +1
      160-199 pts     → defender +2
      200 pts         → defender +3
    """
    if total_points == 0:
        return (3, False)
    if total_points < 40:
        return (2, False)
    if total_points < 80:
        return (1, False)
    if total_points < 120:
        return (0, True)
    if total_points < 160:
        return (-1, True)
    if total_points < 200:
        return (-2, True)
    return (-3, True)


def _advance_level(current: Rank, change: int) -> Rank:
    """Advance a level by change steps, clamping to LEVELS bounds."""
    idx = LEVELS.index(current)
    new_idx = max(0, min(len(LEVELS) - 1, idx + change))
    return LEVELS[new_idx]


def calculate_score(
    defender_points: int,
    bottom_cards: list[Card],
    last_trick_winner_team: int,
    last_trick_play_type: PlayType,
    declarer_team_index: int,
    declarer_team_level: Rank,
    defender_team_level: Rank,
) -> ScoreResult:
    """Calculate the score result for a round.

    Args:
        defender_points: Points collected by defender from tricks.
        bottom_cards: The 8 bottom cards (底牌).
        last_trick_winner_team: Which team won the last trick (0 or 1).
        last_trick_play_type: Play type of the last trick.
        declarer_team_index: Which team was declarer this round (0 or 1).
        declarer_team_level: Current level of the declarer team.
        defender_team_level: Current level of the defender team.

    Returns:
        ScoreResult with level changes, new levels, and point breakdown.
    """
    defender_team_index = 1 - declarer_team_index

    # Calculate bottom card bonus (扣底)
    bp = _bottom_card_points(bottom_cards)
    ambush = last_trick_winner_team == defender_team_index
    multiplier = _AMBUSH_MULTIPLIER.get(last_trick_play_type, 2) if ambush else 0
    bottom_card_bonus = bp * multiplier

    total_points = defender_points + bottom_card_bonus

    # Determine level change
    declarer_change, switch_declarer = _determine_level_change(total_points)

    # Calculate new levels independently for each team
    if declarer_team_index == 0:
        team0_new_level = _advance_level(declarer_team_level, declarer_change)
        team1_new_level = defender_team_level
    else:
        team0_new_level = declarer_team_level
        team1_new_level = _advance_level(declarer_team_level, declarer_change)

    return ScoreResult(
        declarer_level_change=declarer_change,
        switch_declarer=switch_declarer,
        bottom_card_bonus=bottom_card_bonus,
        total_defender_points=total_points,
        team0_new_level=team0_new_level,
        team1_new_level=team1_new_level,
    )


def is_game_over(level: Rank, target_level: Rank = TARGET_LEVEL) -> bool:
    """Check if a team has reached or passed the target level (game over).

    Bug #6 fix: uses >= instead of > (reaching A is sufficient).
    """
    return LEVELS.index(level) >= LEVELS.index(target_level)
